### DISCUSSION TITLE
How to enable detekt with type resolution for subprojects?

### DISCUSSION THREAD

**Discussion by QuinnBast** [detekt/detekt#4137] [Q&A]
Currently I am running detekt in a project with subprojects. Specifically, our project is configured using something [very similar to this](https://madhead.me/posts/detekt-faster/) (code below) where we generate one report for all subprojects.

```kotlin
import io.gitlab.arturbosch.detekt.Detekt

plugins {
    id("io.gitlab.arturbosch.detekt").version("1.18.1")
}

configure<DetektExtension> {
    toolVersion = "1.18.1"
    config = files("${rootProject.projectDir}/detekt-config.yml")
    buildUponDefaultConfig = true
    source = files(rootDir)
    parallel = true
}

tasks.withType<Detekt>().configureEach {
    // Target version of the generated JVM bytecode. It is used for type resolution.
    // I tried adding this but it did not work...
    this.jvmTarget = "1.8"
    exclude("**/build/**") // exclude all generated files in the build directories
    exclude("**/testutils/data/**")
}
```

I cannot figure out how to add Type resolution to our configuration using this setup and any help would be appreciated!

**Comment by QuinnBast** [comment]
Currently experimenting with this configuration that depends on the `jar` task and passes the `build` directory into the classpath:

```kotlin
tasks.withType<Detekt>().configureEach {
    this.jvmTarget = "1.8"
    val originalClassPaths = this.classpath.map { it.path } // Stack overflow if I just pass in the original list...
    dependsOn(tasks.compileJava)
    dependsOn(tasks.inspectClassesForKotlinIC)
    dependsOn(tasks.jar)
    dependsOn(tasks.processResources)

    this.classpath.setFrom(listOf(originalClassPaths, "$rootDir/build/"))

    exclude("**/build/**") // exclude all generated files in the build directories
    exclude("**/testutils/data/**")
}
```

However, I keep getting this error with no idea how to resolve it:

> Execution failed for task ':detekt'.
> Front-end Internal error: Failed to analyze declaration Build_gradle
  File being compiled: (1,1) in \<path\>/build.gradle.kts
  The root cause org.jetbrains.kotlin.resolve.lazy.NoDescriptorForDeclarationException was thrown at: org.jetbrains.kotlin.resolve.lazy.BasicAbsentDescriptorHandler.diagnoseDescriptorNotFound(AbsentDescriptorHandler.kt:18)

This error appears to be a jetbrains issue, however, it is only occurring when I have the `this.classpath.setFrom()` line commented in and include anything that is not the original classpath.

**Comment by cortinico** [comment]
> I cannot figure out how to add Type resolution to our configuration using this setup and any help would be appreciated!

What's the reason why you can't use the `detektMain` and other generated methods for JVM/Android/Multiplatform projects?

**Reply by QuinnBast** [reply]
The way we have detek configured, just runnung `./gradlew detektMain` doesn't not work and does not find errors in our project. I added a wildcard import to my project for testing and `./gradlew detekt` found it and threw an error, but `./gradlew detektMain` said that the build was successful and found no error at all...

**Reply by BraisGabin** [reply]
This is the configuration in the project that I'm working right now and the type solving works like a charm. All this code goes in the root `build.gradle.kts`:

```kotlin
plugins {
    id("io.gitlab.arturbosch.detekt") version "1.18.1"
    ...
}

subprojects {
    apply(plugin = "io.gitlab.arturbosch.detekt")

    detekt {
        baseline = file("baseline.xml")
    }

    tasks.withType<io.gitlab.arturbosch.detekt.Detekt>().configureEach {
        jvmTarget = "11"
    }

    tasks.withType<io.gitlab.arturbosch.detekt.DetektCreateBaselineTask>().configureEach {
        jvmTarget = "11"
    }
}
```

There are other ways to configure it too.
